trieste/type.py

```diff
@@ -12,10 +12,19 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 """This module contains type aliases."""
-from typing import Union
+from typing import Callable, Optional, Tuple, TypeVar, Union
 
 import numpy as np
 import tensorflow as tf
 
 TensorType = Union[np.ndarray, tf.Tensor]
 """Type alias for tensor-like types."""
+
+S = TypeVar("S")
+""" Unbound type variable. """
+
+T = TypeVar("T")
+""" Unbound type variable. """
+
+State = Callable[[Optional[S]], Tuple[S, T]]
+""" A `State` represents a stateful function, with state of type `S` and output of type `T`. """
```

Comment:
tbh i'm not sure it's right to allow `Optional[S]`. I think `S` is more appropriate. I may try to figure this out